src/ConfigParser.cpp

```diff
@@ -18,21 +25,142 @@ static std::string trim(const std::string& str)
+	return segmentCount == 4;
+}
+
+static std::string adjustPath(const std::string& path) 
+{
+	if (!path.empty() && path[0] == '/')
+		return "." + path;
+	return "./" + path;
+}
+
+static bool directoryExists(const std::string& path)
+{
+	struct stat info;
+	if (stat(path.c_str(), &info) != 0)
+		return false;
+	return (info.st_mode & S_IFDIR) != 0;
+}
+
+static bool checkValidDirectory(const std::string& line)
+{
+	std::istringstream iss(line);
+	std::string firstWord;
+	iss >> firstWord;
+
+	if (firstWord == "directory") 
+	{
+		std::string directoryPath;
+		if (iss >> directoryPath) 
+		{
+			std::string adjustedPath = adjustPath(directoryPath);
+			return directoryExists(adjustedPath);
+		}
+	}
+	return true;
+}
+
+static void uniChecker(const std::vector<Server>& servers)
+{
+	std::set<std::string> uniqueNames;
+	std::set<std::string> uniqueHostPortCombos;
+
+	for (const Server& server: servers)
+	{
+		std::string name = server.getName();
+		std::string hostPortCombo = server.getHostIp() + ":" + std::to_string(server.getListenPort());
+
+		if (!uniqueNames.insert(name).second) 
+			throw ConfigurationException("Duplicate name detected.");
+		if (!uniqueHostPortCombos.insert(hostPortCombo).second)
+			throw ConfigurationException("Duplicate IP/Port combination detected.");   
+	}
+}
+
+static void clearMap(vectorMap& vMap)
+{
+	for (vectorMap::iterator it = vMap.begin(); it != vMap.end(); ++it) 
+	{
+		it->second.clear();
+	}
+	vMap.clear();
+}
+
+
+const std::vector<Server>& ConfigParser::getServers() const
+{
+	return servers;
+}
+
+void ConfigParser::checkServer()
+{
+	if (sectionStack.size() != 0)
+		configError("Unclosed block before new server declaration.", lineNumber);
+	currentSection.clear();
+	currentLocation.clear();
+	Server currentServer;
+	servers.push_back(currentServer);
+}
+
+void ConfigParser::checkMain(const std::string& keyword, const std::string& value) 
+{
+	if (servers.empty()) 
+		configError("No server defined for main block.", lineNumber);
+
+	Server& currentServer = servers.back();
+
+	if (keyword.compare(PARSEHOST) == 0) 
+	{
+		if (!validIp(value))
+			configError("Invalid IP address.", lineNumber);
+		currentServer.setHostIp(value);
+	} 
+	else if (keyword.compare(PARSELISTEN) == 0) 
+	{
+		size_t port = std::stol(value);
+		if (port > UINT16_MAX || port < 0 || value.empty())
+			configError("Invalid port.", lineNumber);
+		currentServer.setListenPort(port);
+	} 
+	else if (keyword.compare(PARSENAME) == 0) 
+	{
+		if (value.empty())
+			configError("No server name provided.", lineNumber);
+		currentServer.setName(value);
+	} 
+	else if (keyword.compare(PARSESIZE) == 0) 
+		currentServer.setClientMaxBodySize(std::stol(value));
+	// TODO: check other keywords?
+}
+
 void ConfigParser::parseConfig(const std::string& filename)
 {
 	std::ifstream file(filename);
 	std::string line;
 
 	if (!file.is_open())
-		throw ConfigurationException("Failed to open file.");
+		configError("Failed to open file.", lineNumber);
```

Comment:
I added some checks for the FileHandler here, .bad() and another one, maybe those could be added here too? Also, why the lineNumber here now that we just cant open the file?